common/src/logging.rs

```diff
@@ -29,26 +30,84 @@ pub enum LoggingError {
 
 pub fn init(level: &str, mode: Mode) -> Result<(), LoggingError> {
 	let reload = RELOAD_HANDLE.get_or_try_init(|| {
-		let env_filter = EnvFilter::from_str(level).expect("failed to parse log level");
+		let env_filter = EnvFilter::from_str(level)?;
+
+		match mode {
+			Mode::Default => {
+				let filter = tracing_subscriber::fmt()
+					.with_line_number(true)
+					.with_file(true)
+					.with_env_filter(env_filter)
+					.with_filter_reloading();
 
-		let filter = tracing_subscriber::fmt()
-			.with_line_number(true)
-			.with_file(true)
-			.with_env_filter(env_filter)
-			.with_filter_reloading();
+				let handle = filter.reload_handle();
 
-		let handle = filter.reload_handle();
+				filter.finish().try_init()?;
 
-		match mode {
-			Mode::Default => filter.finish().try_init(),
-			Mode::Json => filter.json().finish().try_init(),
-			Mode::Pretty => filter.pretty().finish().try_init(),
-			Mode::Compact => filter.compact().finish().try_init(),
+				Ok::<_, LoggingError>(Box::new(move |level: &str| {
+					let level = EnvFilter::from_str(level)?;
+					handle.reload(level)?;
+					Ok(())
+				}) as ReloadHandle)
+			}
+			Mode::Json => {
+				let filter = tracing_subscriber::fmt()
+					.json()
+					.with_line_number(true)
+					.with_file(true)
+					.with_env_filter(env_filter)
+					.with_filter_reloading();
+
+				let handle = filter.reload_handle();
+
+				filter.finish().try_init()?;
+
+				Ok(Box::new(move |level: &str| {
+					let level = EnvFilter::from_str(level)?;
+					handle.reload(level)?;
+					Ok(())
+				}) as ReloadHandle)
+			}
+			Mode::Pretty => {
+				let filter = tracing_subscriber::fmt()
+					.pretty()
+					.with_line_number(true)
+					.with_file(true)
+					.with_env_filter(env_filter)
+					.with_filter_reloading();
+
+				let handle = filter.reload_handle();
+
+				filter.finish().try_init()?;
+
+				Ok(Box::new(move |level: &str| {
+					let level = EnvFilter::from_str(level)?;
+					handle.reload(level)?;
+					Ok(())
+				}) as ReloadHandle)
+			}
+			Mode::Compact => {
+				let filter = tracing_subscriber::fmt()
+					.compact()
+					.with_line_number(true)
+					.with_file(true)
+					.with_env_filter(env_filter)
+					.with_filter_reloading();
+
+				let handle = filter.reload_handle();
+
+				filter.finish().try_init()?;
+
+				Ok(Box::new(move |level: &str| {
+					let level = EnvFilter::from_str(level)?;
+					handle.reload(level)?;
+					Ok(())
+				}) as ReloadHandle)
+			}
 		}
```

Comment:
Why is there so much code duplication here? Can't this be solved in a more elegant way?

Reply:
Unfortunately no, tracing has layers, by changing `.json()` or `.compact()` or `.pretty()` it changes a root layer meaning the types are different on the match branches.

Reply:
Not without a macro.